In [1]:
import pandas as pd

# Data Loading and Observation
Loading and observing the data entries in the datasets from Olist.

In [2]:
# Load the customers dataset
customers_df = pd.read_csv(
    "../data/01_raw/olist_customers_dataset.csv",
    converters={
        'customer_zip_code_prefix': str # Convert ZIP codes to strings
    }
)
customers_df

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

      customer_zip_code_prefix          customer_city customer_state  
0                        14409                 franca             SP  
1                        09790  sao bernardo do campo             SP  
2                        01151              sao paulo             SP  
3                        08775        mogi das cruzes             SP  
4                        13056               campinas             SP  
...                        ...                    ...            ...  
99436                    03937              sao paulo             SP  
99437                    06764        taboao da serra             SP  
99438                    60115              fortaleza             CE  
99439                    92120                 canoas             RS  
99440                    06703                  cotia             SP  

[99441 rows x 5 columns]

In [3]:
# Load the geolocation dataset
geolocation_df = pd.read_csv(
    "../data/01_raw/olist_geolocation_dataset.csv",
    converters={
        'geolocation_zip_code_prefix': str # Convert ZIP codes to strings
    }
)
geolocation_df

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                             01037       -23.545621       -46.639292   
1                             01046       -23.546081       -46.644820   
2                             01046       -23.546129       -46.642951   
3                             01041       -23.544392       -46.639499   
4                             01035       -23.541578       -46.641607   
...                             ...              ...              ...   
1000158                       99950       -28.068639       -52.010705   
1000159                       99900       -27.877125       -52.224882   
1000160                       99950       -28.071855       -52.014716   
1000161                       99980       -28.388932       -51.846871   
1000162                       99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
0              sao paulo                SP  
1              sao paulo                SP  
2              sao paulo                SP  
3              sao paulo                SP  
4              sao paulo                SP  
...                  ...               ...  
1000158         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000161  david canabarro                RS  
1000162         tapejara                RS  

[1000163 rows x 5 columns]

In [4]:
# Load the order items dataset
items_df = pd.read_csv("../data/01_raw/olist_order_items_dataset.csv")
items_df

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  
0       2017-09-19 09:45:35   58.90          13.29  
1       2017-05-03 11:05:13  239.90          19.93  
2       2018-01-18 14:48:30  199.00          17.87  
3       2018-08-15 10:10:18   12.99          12.79  
4       2017-02-13 13:57:51  199.90          18.14  
...                     ...     ...            ...  
112645  2018-05-02 04:11:01  299.99          43.41  
112646  2018-07-20 04:31:48  350.00          36.53  
112647  2017-10-30 17:14:25   99.90          16.95  
112648  2017-08-21 00:04:32   55.99           8.72  
112649  2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

In [5]:
# Load the payments dataset
payments_df = pd.read_csv("../data/01_raw/olist_order_payments_dataset.csv")
payments_df

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
0                          8          99.33  
1                          1          24.39  
2                          1          65.71  
3                          8         107.78  
4                          2         128.45  
...                      ...            ...  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58  

[103886 rows x 5 columns]

In [6]:
# Load the reviews dataset
reviews_df = pd.read_csv("../data/01_raw/olist_order_reviews_dataset.csv")
reviews_df

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [7]:
# Load the orders dataset
orders_df = pd.read_csv("../data/01_raw/olist_orders_dataset.csv")
orders_df

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 8 columns]

In [8]:
# Load the products dataset
products_df = pd.read_csv("../data/01_raw/olist_products_dataset.csv")
products_df

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32951 rows x 9 columns]

In [9]:
# Load the sellers dataset
sellers_df = pd.read_csv(
    "../data/01_raw/olist_sellers_dataset.csv",
    converters={
        'seller_zip_code_prefix': str # Convert ZIP codes to strings
    }
)
sellers_df

seller_id seller_zip_code_prefix  \
0     3442f8959a84dea7ee197c632cb2df15                  13023   
1     d1b65fc7debc3361ea86b5f14c68d2e2                  13844   
2     ce3ad9de960102d0677a81f5d0bb7b2d                  20031   
3     c0f3eea2e14555b6faeea3dd58c1b1c3                  04195   
4     51a04a8a6bdcb23deccc82b0b80742cf                  12914   
...                                ...                    ...   
3090  98dddbc4601dd4443ca174359b237166                  87111   
3091  f8201cab383e484733266d1906e2fdfa                  88137   
3092  74871d19219c7d518d0090283e03c137                  04650   
3093  e603cf3fec55f8697c9059638d6c8eb5                  96080   
3094  9e25199f6ef7e7c347120ff175652c3b                  12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 4 columns]

In [10]:
# Load the product category name translation dataset
translations_df = pd.read_csv("../data/01_raw/product_category_name_translation.csv")
translations_df

product_category_name product_category_name_english
0                    beleza_saude                 health_beauty
1          informatica_acessorios         computers_accessories
2                      automotivo                          auto
3                 cama_mesa_banho                bed_bath_table
4                moveis_decoracao               furniture_decor
..                            ...                           ...
66                         flores                       flowers
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services

[71 rows x 2 columns]

# Data Quality Check
Identifies any data quality issues.

## Overview
High level overview of the structure of the datasets.

In [11]:
# Obtain a high level overview of the data in the customers dataset
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  object
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: object(5)
memory usage: 3.8+ MB


In [12]:
# Obtain a high level overview of the data in the geolocation dataset
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  object 
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), object(3)
memory usage: 38.2+ MB


In [13]:
# Obtain a high level overview of the data in the items dataset
items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [14]:
# Obtain a high level overview of the data in the payments dataset
payments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [15]:
# Obtain a high level overview of the data in the reviews dataset
reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [16]:
# Obtain a high level overview of the data in the orders dataset
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [17]:
# Obtain a high level overview of the data in the products dataset
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [18]:
# Obtain a high level overview of the data in the sellers dataset
sellers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   object
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: object(4)
memory usage: 96.8+ KB


In [19]:
# Obtain a high level overview of the data in the translations dataset
translations_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


Notes:
- All the datasets have different number of rows, suggesting that there is not a one-to-one relationship between any of the datasets.
- The reviews dataset contains a significant number of missing comment titles and messages. review_comment_title has only 11568 non-null values and review_comment_message only has 40977 non-null values out of the 99 224 rows.
- The orders dataset contains a slight amount of missing values in the approval date, delivered carrier date, and delivered customer date.
- The products dataset also contains a small amount of missing values in all columns except product_id.

Datasets With Missing Entries:
- reviews
- orders
- products

## Distributions
Obtain a high-level overview of the distributions of data in the datasets.

In [20]:
# Observe the data distributions in the customers dataset
customers_df.describe()

customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     274fa6071e5e17fe303b9748641082c8  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746

Notes:
- All customer_id entries are unique as count and unique are equal (99441).
- Not all customer_unique_id entries are unique as count and unique are not equal (99441 and 96096 respectively).
- Not all customer_zip_code_prefix are unique, up to 142 customers are living in the same zip code area.
- Sao paulo citizens are the highest proportion of Olist's customers, with 15 540 customers from Sao Paulo.
- SP state citizens are the highest proportion of Olist's customers, with 41 746 customers from SP.

In [21]:
# Set float display format to prevent scientific notation
pd.options.display.float_format = '{:.5f}'.format

# Observe the data distributions in the geolocations dataset
geolocation_df.describe(include=['float64', 'object']).round(2)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
count                      1000163    1000163.00000    1000163.00000   
unique                       19015              NaN              NaN   
top                          24220              NaN              NaN   
freq                          1146              NaN              NaN   
mean                           NaN        -21.18000        -46.39000   
std                            NaN          5.72000          4.27000   
min                            NaN        -36.61000       -101.47000   
25%                            NaN        -23.60000        -48.57000   
50%                            NaN        -22.92000        -46.64000   
75%                            NaN        -19.98000        -43.77000   
max                            NaN         45.07000        121.11000   

       geolocation_city geolocation_state  
count           1000163           1000163  
unique             8011                27  
top           sao paulo                SP  
freq             135800            404268  
mean                NaN               NaN  
std                 NaN               NaN  
min                 NaN               NaN  
25%                 NaN               NaN  
50%                 NaN               NaN  
75%                 NaN               NaN  
max                 NaN               NaN

In [22]:
# Reset the display format back to the default
pd.reset_option('display.float_format')

Notes:
- The range of Brazils latitude is between -33 and 5. The range of Brazils longitude is between -34 and -73.
- However, the max geolocation latitude is 45.07000. Furthermore, the geolocation longitude range is between -101.47000	and 121.11000, suggesting that there are entries which are outside Brazil's longitude and latitude. This suggests that the geolocation entries contain entries which are not located in Brazil. Investigated in Geolocation Check.

In [23]:
# Observe the data distributions in the items dataset
items_df.describe(include=['float64', 'int64', 'object'])

order_id  order_item_id  \
count                             112650  112650.000000   
unique                             98666            NaN   
top     8272b63d03f5f79c56e9e4120aec44ef            NaN   
freq                                  21            NaN   
mean                                 NaN       1.197834   
std                                  NaN       0.705124   
min                                  NaN       1.000000   
25%                                  NaN       1.000000   
50%                                  NaN       1.000000   
75%                                  NaN       1.000000   
max                                  NaN      21.000000   

                              product_id                         seller_id  \
count                             112650                            112650   
unique                             32951                              3095   
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b47992c3666cc44a7e94c0   
freq                                 527                              2033   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        shipping_limit_date          price  freight_value  
count                112650  112650.000000  112650.000000  
unique                93318            NaN            NaN  
top     2018-03-01 02:50:48            NaN            NaN  
freq                     21            NaN            NaN  
mean                    NaN     120.653739      19.990320  
std                     NaN     183.633928      15.806405  
min                     NaN       0.850000       0.000000  
25%                     NaN      39.900000      13.080000  
50%                     NaN      74.990000      16.260000  
75%                     NaN     134.900000      21.150000  
max                     NaN    6735.000000     409.680000

Notes:
- The minimum freight_value is 0 suggesting that delivery is free for some products.
- One product was purchased in bulk, purchasing 21 of the same product in the same order. This was shown as the max value in unique order_item_id is 21, suggesting that there was a specific product that was purchased 21 times in the same order. This may be a n important product that needs to be investigated in the EDA.

In [24]:
# Observe the data distributions in the payments dataset
payments_df.describe(include=['float64', 'int64', 'object'])

order_id  payment_sequential payment_type  \
count                             103886       103886.000000       103886   
unique                             99440                 NaN            5   
top     fa65dad1b0e818e3ccc5cb0e39231352                 NaN  credit_card   
freq                                  29                 NaN        76795   
mean                                 NaN            1.092679          NaN   
std                                  NaN            0.706584          NaN   
min                                  NaN            1.000000          NaN   
25%                                  NaN            1.000000          NaN   
50%                                  NaN            1.000000          NaN   
75%                                  NaN            1.000000          NaN   
max                                  NaN           29.000000          NaN   

        payment_installments  payment_value  
count          103886.000000  103886.000000  
unique                   NaN            NaN  
top                      NaN            NaN  
freq                     NaN            NaN  
mean                2.853349     154.100380  
std                 2.687051     217.494064  
min                 0.000000       0.000000  
25%                 1.000000      56.790000  
50%                 1.000000     100.000000  
75%                 4.000000     171.837500  
max                24.000000   13664.080000

Notes:
- There are customers who want to pay with up to 29 payment methods which is an outlier as it does not seem plausible to pay an order using a large variety of payment methods. This will be investigated further in Payment Sequential Check.
- There are orders which can be paid by up to 24 payment installments. This is shown as the max payment_sequential is 24.
- Some orders are free. The min payment_value is 0 suggesting that some orders are completely free.

In [25]:
# Observe the data distributions in the reviews dataset
reviews_df.describe(include=['float64', 'int64', 'object'])

review_id                          order_id  \
count                              99224                             99224   
unique                             98410                             98673   
top     08528f70f579f0c830189efc523d2182  df56136b8031ecd28e200bb18e6ddb2e   
freq                                   3                                 3   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        review_score review_comment_title review_comment_message  \
count   99224.000000                11568                  40977   
unique           NaN                 4527                  36159   
top              NaN            Recomendo              Muito bom   
freq             NaN                  423                    230   
mean        4.086421                  NaN                    NaN   
std         1.347579                  NaN                    NaN   
min         1.000000                  NaN                    NaN   
25%         4.000000                  NaN                    NaN   
50%         5.000000                  NaN                    NaN   
75%         5.000000                  NaN                    NaN   
max         5.000000                  NaN                    NaN   

       review_creation_date review_answer_timestamp  
count                 99224                   99224  
unique                  636                   98248  
top     2017-12-19 00:00:00     2017-06-15 23:21:05  
freq                    463                       4  
mean                    NaN                     NaN  
std                     NaN                     NaN  
min                     NaN                     NaN  
25%                     NaN                     NaN  
50%                     NaN                     NaN  
75%                     NaN                     NaN  
max                     NaN                     NaN

Notes:
- Generally, the olist's orders are completed with good reviews. The median review score is 5/5, and the 25% percentile of the review score is 4/5 which is still decent.
- Majority of the reviews do not have a title or a message. The count of review_comment_title and review_comment_message, 11568 and 40977 respectively, are significantly lower than the number of reviews in the dataset (99224). 

In [26]:
# Observe the data distributions in the orders dataset
orders_df.describe(include=['float64', 'int64', 'object'])

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-08-02 12:05:26  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 19:36:48   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

Notes:
- This dataset identifies the state of the order in the supply chain.
- The missing values in the order_approved_at, order_delivered_carrier_date,order_delivered_customer_date fields is likely due to the status of the order, having not passed certain parts of the supply chain yet. 

In [27]:
# Observe the data distributions in the products dataset
products_df.describe(include=['float64', 'int64', 'object'])

product_id product_category_name  \
count                              32951                 32341   
unique                             32951                    73   
top     106392145fca363410d287a815be6de4       cama_mesa_banho   
freq                                   1                  3029   
mean                                 NaN                   NaN   
std                                  NaN                   NaN   
min                                  NaN                   NaN   
25%                                  NaN                   NaN   
50%                                  NaN                   NaN   
75%                                  NaN                   NaN   
max                                  NaN                   NaN   

        product_name_lenght  product_description_lenght  product_photos_qty  \
count          32341.000000                32341.000000        32341.000000   
unique                  NaN                         NaN                 NaN   
top                     NaN                         NaN                 NaN   
freq                    NaN                         NaN                 NaN   
mean              48.476949                  771.495285            2.188986   
std               10.245741                  635.115225            1.736766   
min                5.000000                    4.000000            1.000000   
25%               42.000000                  339.000000            1.000000   
50%               51.000000                  595.000000            1.000000   
75%               57.000000                  972.000000            3.000000   
max               76.000000                 3992.000000           20.000000   

        product_weight_g  product_length_cm  product_height_cm  \
count       32949.000000       32949.000000       32949.000000   
unique               NaN                NaN                NaN   
top                  NaN                NaN                NaN   
freq                 NaN                NaN                NaN   
mean         2276.472488          30.815078          16.937661   
std          4282.038731          16.914458          13.637554   
min             0.000000           7.000000           2.000000   
25%           300.000000          18.000000           8.000000   
50%           700.000000          25.000000          13.000000   
75%          1900.000000          38.000000          21.000000   
max         40425.000000         105.000000         105.000000   

        product_width_cm  
count       32949.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean           23.196728  
std            12.079047  
min             6.000000  
25%            15.000000  
50%            20.000000  
75%            30.000000  
max           118.000000

Notes:
- There is a product that weights 0g. This outlier is potentially a data entry error as products should have a mass. Investigated in Product Weight Check.
- There is an anomalously long product_description_lenght entry with 3992 characters. This may be using many different tags to push the product in search results. This may be investigated in the EDA.
- There are products which have anomalously high product_photos_qty (20). More pictures may improve customer satisfaction with the product ordered as they may be more aware of what the product will look like. This may be investigated in the EDA.

In [28]:
# Observe the data distributions in the sellers dataset
sellers_df.describe(include=['float64', 'int64', 'object'])

seller_id seller_zip_code_prefix seller_city  \
count                               3095                   3095        3095   
unique                              3095                   2246         611   
top     9e25199f6ef7e7c347120ff175652c3b                  14940   sao paulo   
freq                                   1                     49         694   

       seller_state  
count          3095  
unique           23  
top              SP  
freq           1849

Notes:
- Nothing noteworthy

In [29]:
# Observe the data distributions in the sellers dataset
translations_df.describe(include=['float64', 'int64', 'object'])

product_category_name product_category_name_english
count                     71                            71
unique                    71                            71
top             beleza_saude                 health_beauty
freq                       1                             1

Notes:
- Nothing noteworthy
- Can be useful to understand the categories of the products in English.

## Null Check
Investigates the missing values identified in the Overview section.

In [67]:
# Calculates the percentage of null entries in each column i.e. number of null entries / number of entries in the column
def calculate_null_proportion_per_column(df):
    num_null_entries_per_column = df.isnull().sum()
    total_rows = df.shape[0]
    null_proportion = num_null_entries_per_column/total_rows *100
    print("% of null entries per column:")
    return null_proportion.round(2)

# Calculates the number and percentage of rows with at least one null entry
def count_rows_with_nulls(df):
    # Count rows with at least one null
    num_rows_with_nulls = df.isnull().any(axis=1).sum()
    total_rows = df.shape[0]
    percent_rows_with_null = (num_rows_with_nulls / total_rows) * 100  # Calculate percentage

    # Display results
    print(f"No. rows with at least one null value: {num_rows_with_nulls}")
    print(f"% of rows with at least one null value: {percent_rows_with_null:.2f}%")

The following datasets have missing values and will be the only ones investigated in this section
- Reviews
- Orders
- Products

The other datasets do not have null values in them and will not need to be investigated for null values.

### Reviews
Investigates the missing values in the reviews dataset

In [68]:
# Investigate the proportion of null entries in each data field of the reviews dataset
calculate_null_proportion_per_column(reviews_df)

% of null entries per column:


review_id                   0.00
order_id                    0.00
review_score                0.00
review_comment_title       88.34
review_comment_message     58.70
review_creation_date        0.00
review_answer_timestamp     0.00
dtype: float64

In [69]:
# Identify the number of rows with at least 1 null value in the reviews dataset
count_rows_with_nulls(reviews_df)

No. rows with at least one null value: 89385
% of rows with at least one null value: 90.08%


In [70]:
# Identify the rows of data with missing comment titles
missing_comment_titles_df = reviews_df[reviews_df['review_comment_title'].isnull()]
missing_comment_titles_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [34]:
# Identify the rows of data with missing comment messages
missing_comment_message_df = reviews_df[reviews_df['review_comment_message'].isnull()]
missing_comment_message_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   

   review_score review_comment_title review_comment_message  \
0             4                  NaN                    NaN   
1             5                  NaN                    NaN   
2             5                  NaN                    NaN   
5             1                  NaN                    NaN   
6             5                  NaN                    NaN   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13  
2  2018-02-17 00:00:00     2018-02-18 14:36:24  
5  2018-04-13 00:00:00     2018-04-16 00:39:37  
6  2017-07-16 00:00:00     2017-07-18 19:30:34

In [35]:
# Identify the rows of data with missing comment messages but with a comment title
missing_message_existing_title_df = reviews_df[
    reviews_df['review_comment_message'].isnull() & reviews_df['review_comment_title'].notnull()
]
missing_message_existing_title_df.head()

review_id                          order_id  \
53   1edd1a85161468a77e5fcb165be9b294  d8e17cc85323061493b16b68a130bcff   
104  352609f0f2d34ce5d3b8de817ee948d5  96b947c986b42f213c9e6b8ccccedb89   
220  a89a2b0c2014bc49a0f87f1eedb7b269  22b1eef278d4416c1e867a55d1c39fea   
225  cdc5c81bd9377ccbf9cb95f05648c31b  18956b40b774d6384c8539cd61784bd7   
495  b50001aa0bbd7557c4ad80eee167a2a1  daece90b61801c766fe21e0922574566   

     review_score       review_comment_title review_comment_message  \
53              5                      Ótimo                    NaN   
104             3      Não recebi o produto.                    NaN   
220             5  Entrega antes do previsto                    NaN   
225             5                  Recomendo                    NaN   
495             5                     Ótimo                     NaN   

    review_creation_date review_answer_timestamp  
53   2018-06-02 00:00:00     2018-06-02 23:15:37  
104  2018-08-25 00:00:00     2018-08-25 20:53:41  
220  2018-05-26 00:00:00     2018-05-28 22:46:27  
225  2018-06-22 00:00:00     2018-06-24 23:07:08  
495  2018-05-05 00:00:00     2018-05-07 17:12:36

Notes:
- There is a substantially large proportion of missing entries for review_comment_title (88.34%). This is not anomalous as the reviewer likely wanted to leave a brief review without any need for a title. The system for making the review likely did not require a title.
- There is also a significant proportion of missing entries for review_comment_message (55.70%). This is not anomalous as the reviewer likely did not want to elaborate on details of the product in the review. The system for making the review likely did not require the user to write a review message too.
- These rows should not be dropped as they are not data entry errors - the reviewer just did not want to enter a title or message.
- Furthermore, the rows with null values should also not be dropped because 90.08% of the rows have at least 1 null value, this proportion of rows is too significant to be dropped anyway.

### Orders
Investigates the missing values in the orders dataset

In [36]:
# Investigate the proportion of null entries in each data field of the orders dataset
calculate_null_proportion_per_column(orders_df)

% of null entries per column:


order_id                         0.00
customer_id                      0.00
order_status                     0.00
order_purchase_timestamp         0.00
order_approved_at                0.16
order_delivered_carrier_date     1.79
order_delivered_customer_date    2.98
order_estimated_delivery_date    0.00
dtype: float64

In [37]:
# Identify the number of rows with at least 1 null value in the orders dataset
count_rows_with_nulls(orders_df)

No. rows with at least one null value: 2980
% of rows with at least one null value: 3.00%


In [38]:
# Identify rows with missing order approved dates
missing_approved_df = orders_df[orders_df['order_approved_at'].isnull()]
missing_approved_df.head()

order_id                       customer_id  \
1130  00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801  ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
1868  df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2029  8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2161  7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   

     order_status order_purchase_timestamp order_approved_at  \
1130     canceled      2018-08-28 15:26:39               NaN   
1801     canceled      2018-09-20 13:54:16               NaN   
1868     canceled      2017-03-04 12:14:30               NaN   
2029     canceled      2018-08-29 16:27:49               NaN   
2161     canceled      2017-05-01 16:12:39               NaN   

     order_delivered_carrier_date order_delivered_customer_date  \
1130                          NaN                           NaN   
1801                          NaN                           NaN   
1868                          NaN                           NaN   
2029                          NaN                           NaN   
2161                          NaN                           NaN   

     order_estimated_delivery_date  
1130           2018-09-12 00:00:00  
1801           2018-10-17 00:00:00  
1868           2017-04-10 00:00:00  
2029           2018-09-13 00:00:00  
2161           2017-05-30 00:00:00

In [39]:
# Identify the types of order statuses associated with missing order_approved_at data
missing_approved_df['order_status'].value_counts()

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

In [40]:
# Observe the rows whose order statuses are delivered or created which are also missing order_approved_at data
incorrect_approved_df = missing_approved_df[missing_approved_df["order_status"]=='delivered']
incorrect_approved_df

order_id                       customer_id  \
5323   e04abd8149ef81b95221e88f6ed9ab6a  2127dc6603ac33544953ef05ec155771   
16567  8a9adc69528e1001fc68dd0aaebbb54a  4c1ccc74e00993733742a3c786dc3c1f   
19031  7013bcfc1c97fe719a7b5e05e61c12db  2941af76d38100e0f8740a374f1a5dc3   
22663  5cf925b116421afa85ee25e99b4c34fb  29c35fc91fc13fb5073c8f30505d860d   
23156  12a95a3c06dbaec84bcfb0e2da5d228a  1e101e0daffaddce8159d25a8e53f2b2   
26800  c1d4211b3dae76144deccd6c74144a88  684cb238dc5b5d6366244e0e0776b450   
38290  d69e5d356402adc8cf17e08b5033acfb  68d081753ad4fe22fc4d410a9eb1ca01   
39334  d77031d6a3c8a52f019764e68f211c69  0bf35cac6cc7327065da879e2d90fae8   
48401  7002a78c79c519ac54022d4f8a65e6e8  d5de688c321096d15508faae67a27051   
61743  2eecb0d85f281280f79fa00f9cec1a95  a3d3c38e58b9d2dfb9207cab690b6310   
63052  51eb2eebd5d76a24625b31c33dd41449  07a2a7e0f63fd8cb757ed77d4245623c   
67697  88083e8f64d95b932164187484d90212  f67cd1a215aae2a1074638bbd35a223a   
72407  3c0b8706b065f9919d0505d3b3343881  d85919cb3c0529589c6fa617f5f43281   
84999  2babbb4b15e6d2dfe95e2de765c97bce  74bebaf46603f9340e3b50c6b086f992   

      order_status order_purchase_timestamp order_approved_at  \
5323     delivered      2017-02-18 14:40:00               NaN   
16567    delivered      2017-02-18 12:45:31               NaN   
19031    delivered      2017-02-18 13:29:47               NaN   
22663    delivered      2017-02-18 16:48:35               NaN   
23156    delivered      2017-02-17 13:05:55               NaN   
26800    delivered      2017-01-19 12:48:08               NaN   
38290    delivered      2017-02-19 01:28:47               NaN   
39334    delivered      2017-02-18 11:04:19               NaN   
48401    delivered      2017-01-19 22:26:59               NaN   
61743    delivered      2017-02-17 17:21:55               NaN   
63052    delivered      2017-02-18 15:52:27               NaN   
67697    delivered      2017-02-18 22:49:19               NaN   
72407    delivered      2017-02-17 15:53:27               NaN   
84999    delivered      2017-02-18 17:15:03               NaN   

      order_delivered_carrier_date order_delivered_customer_date  \
5323           2017-02-23 12:04:47           2017-03-01 13:25:33   
16567          2017-02-23 09:01:52           2017-03-02 10:05:06   
19031          2017-02-22 16:25:25           2017-03-01 08:07:38   
22663          2017-02-22 11:23:10           2017-03-09 07:28:47   
23156          2017-02-22 11:23:11           2017-03-02 11:09:19   
26800          2017-01-25 14:56:50           2017-01-30 18:16:01   
38290          2017-02-23 03:11:48           2017-03-02 03:41:58   
39334          2017-02-23 07:23:36           2017-03-02 16:15:23   
48401          2017-01-27 11:08:05           2017-02-06 14:22:19   
61743          2017-02-22 11:42:51           2017-03-03 12:16:03   
63052          2017-02-23 03:09:14           2017-03-07 13:57:47   
67697          2017-02-22 11:31:06           2017-03-02 12:06:06   
72407          2017-02-22 11:31:30           2017-03-03 11:47:47   
84999          2017-02-22 11:23:11           2017-03-03 18:43:43   

      order_estimated_delivery_date  
5323            2017-03-17 00:00:00  
16567           2017-03-21 00:00:00  
19031           2017-03-17 00:00:00  
22663           2017-03-31 00:00:00  
23156           2017-03-20 00:00:00  
26800           2017-03-01 00:00:00  
38290           2017-03-27 00:00:00  
39334           2017-03-22 00:00:00  
48401           2017-03-16 00:00:00  
61743           2017-03-20 00:00:00  
63052           2017-03-29 00:00:00  
67697           2017-03-21 00:00:00  
72407           2017-03-23 00:00:00  
84999           2017-03-31 00:00:00

In [41]:
# Display the number of rows with the erroneous missing order_approved_at data
erroneous_entries_count = incorrect_approved_df.shape[0]
print("Number of erroneous missing entries:", erroneous_entries_count)

Number of erroneous missing entries: 14


These entries are erroneous as they are missing the order approved date. These entries suggest that the orders were already approved and that the products were already delivered. These entries should have an approved date as the order should have already passed the approval stage by the time it was delivered.

In [42]:
# Identify rows with missing order_delivered_carrier_date entries
missing_carrier_df = orders_df[orders_df['order_delivered_carrier_date'].isnull()]
missing_carrier_df

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
266    8e24261a7e58791d10cb1bf9da94df5c  64a254d30eed42cd0e6c36dddb88adf0   
324    d3c8851a6651eeff2f73b0e011ac45d0  957f8e082185574de25992dc659ebbc0   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp    order_approved_at  \
6         invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
103       invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
266    unavailable      2017-11-16 15:09:28  2017-11-16 15:26:57   
324     processing      2016-10-05 22:44:13  2016-10-06 15:51:05   
...            ...                      ...                  ...   
99283     canceled      2018-08-31 16:13:44                  NaN   
99313   processing      2018-01-05 23:07:24  2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                  NaN   
99348  unavailable      2017-08-23 16:28:04  2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03  2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaN                           NaN   
103                            NaN                           NaN   
128                            NaN                           NaN   
266                            NaN                           NaN   
324                            NaN                           NaN   
...                            ...                           ...   
99283                          NaN                           NaN   
99313                          NaN                           NaN   
99347                          NaN                           NaN   
99348                          NaN                           NaN   
99415                          NaN                           NaN   

      order_estimated_delivery_date  
6               2017-05-09 00:00:00  
103             2018-08-21 00:00:00  
128             2017-10-03 00:00:00  
266             2017-12-05 00:00:00  
324             2016-12-09 00:00:00  
...                             ...  
99283           2018-10-01 00:00:00  
99313           2018-02-06 00:00:00  
99347           2018-09-27 00:00:00  
99348           2017-09-15 00:00:00  
99415           2017-10-23 00:00:00  

[1783 rows x 8 columns]

In [43]:
# Identify the types of order statuses associated with missing order_delivered_carrier_date data
missing_carrier_df['order_status'].value_counts()

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

In [44]:
# Observe the entries with the 'delivered' status
erroneous_carrier_df = missing_carrier_df[missing_carrier_df['order_status']=='delivered']
erroneous_carrier_df

order_id                       customer_id  \
73222  2aa91108853cecb43c84a5dc5b277475  afeb16c7f46396c0ed54acb45ccaaa40   
92643  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   

      order_status order_purchase_timestamp    order_approved_at  \
73222    delivered      2017-09-29 08:52:58  2017-09-29 09:07:16   
92643    delivered      2017-05-25 23:22:43  2017-05-25 23:30:16   

      order_delivered_carrier_date order_delivered_customer_date  \
73222                          NaN           2017-11-20 19:44:47   
92643                          NaN                           NaN   

      order_estimated_delivery_date  
73222           2017-11-14 00:00:00  
92643           2017-06-23 00:00:00

These 2 data entries are erroneous as they should have the order_delivered_carrier_date values as they were already delivered to the customer by the carrier, and therefore the carrier should have received the product and the entry of the order_delivered_carrier_date should have been filled. Hence, these 2 data entries are erroneous for missing the values.

In [45]:
# Identify rows with missing order_delivered_customer_date entries
missing_delivered_customer_df = orders_df[orders_df['order_delivered_customer_date'].isnull()]
missing_delivered_customer_df

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp    order_approved_at  \
6         invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
44         shipped      2018-06-04 16:44:48  2018-06-05 04:31:18   
103       invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
154        shipped      2018-01-10 11:33:07  2018-01-11 02:32:30   
...            ...                      ...                  ...   
99283     canceled      2018-08-31 16:13:44                  NaN   
99313   processing      2018-01-05 23:07:24  2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                  NaN   
99348  unavailable      2017-08-23 16:28:04  2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03  2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaN                           NaN   
44             2018-06-05 14:32:00                           NaN   
103                            NaN                           NaN   
128                            NaN                           NaN   
154            2018-01-11 19:39:23                           NaN   
...                            ...                           ...   
99283                          NaN                           NaN   
99313                          NaN                           NaN   
99347                          NaN                           NaN   
99348                          NaN                           NaN   
99415                          NaN                           NaN   

      order_estimated_delivery_date  
6               2017-05-09 00:00:00  
44              2018-06-28 00:00:00  
103             2018-08-21 00:00:00  
128             2017-10-03 00:00:00  
154             2018-02-07 00:00:00  
...                             ...  
99283           2018-10-01 00:00:00  
99313           2018-02-06 00:00:00  
99347           2018-09-27 00:00:00  
99348           2017-09-15 00:00:00  
99415           2017-10-23 00:00:00  

[2965 rows x 8 columns]

In [46]:
# Identify the types of order statuses associated with missing order_delivered_customer_date data
missing_delivered_customer_df['order_status'].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

In [47]:
# Observe the entries with the 'delivered' status
erroneous_delivered_customer_df = missing_delivered_customer_df[missing_delivered_customer_df['order_status']=='delivered']
erroneous_delivered_customer_df

order_id                       customer_id  \
3002   2d1e2d5bf4dc7227b3bfebb81328c15f  ec05a6d8558c6455f0cbbd8a420ad34f   
20618  f5dd62b788049ad9fc0526e3ad11a097  5e89028e024b381dc84a13a3570decb4   
43834  2ebdfc4f15f23b91474edf87475f108e  29f0540231702fda0cfdee0a310f11aa   
79263  e69f75a717d64fc5ecdfae42b2e8e086  cfda40ca8dd0a5d486a9635b611b398a   
82868  0d3268bad9b086af767785e3f0fc0133  4f1d63d35fb7c8999853b2699f5c7649   
92643  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   
97647  ab7c89dc1bf4a1ead9d6ec1ec8968a84  dd1b84a7286eb4524d52af4256c0ba24   
98038  20edc82cf5400ce95e1afacc25798b31  28c37425f1127d887d7337f284080a0f   

      order_status order_purchase_timestamp    order_approved_at  \
3002     delivered      2017-11-28 17:44:07  2017-11-28 17:56:40   
20618    delivered      2018-06-20 06:58:43  2018-06-20 07:19:05   
43834    delivered      2018-07-01 17:05:11  2018-07-01 17:15:12   
79263    delivered      2018-07-01 22:05:55  2018-07-01 22:15:14   
82868    delivered      2018-07-01 21:14:02  2018-07-01 21:29:54   
92643    delivered      2017-05-25 23:22:43  2017-05-25 23:30:16   
97647    delivered      2018-06-08 12:09:39  2018-06-08 12:36:39   
98038    delivered      2018-06-27 16:09:12  2018-06-27 16:29:30   

      order_delivered_carrier_date order_delivered_customer_date  \
3002           2017-11-30 18:12:23                           NaN   
20618          2018-06-25 08:05:00                           NaN   
43834          2018-07-03 13:57:00                           NaN   
79263          2018-07-03 13:57:00                           NaN   
82868          2018-07-03 09:28:00                           NaN   
92643                          NaN                           NaN   
97647          2018-06-12 14:10:00                           NaN   
98038          2018-07-03 19:26:00                           NaN   

      order_estimated_delivery_date  
3002            2017-12-18 00:00:00  
20618           2018-07-16 00:00:00  
43834           2018-07-30 00:00:00  
79263           2018-07-30 00:00:00  
82868           2018-07-24 00:00:00  
92643           2017-06-23 00:00:00  
97647           2018-06-26 00:00:00  
98038           2018-07-19 00:00:00

These 8 data entries are erroneous as they should have the order_delivered_carrier_date values as they were already delivered to the customer as shown by the order status being 'delivered'. Despite this, the 'order_delivered_customer_date' value is missing, suggesting that these data entries are erroneous.

order_approved_at:
- There is a very small proportion of rows with missing order_approved_at dates (0.16%). 
- 155 of these rows are valid as the orders were either created or canceled and therefore not approved and hence will not have an approved date. 
- However, there are 14 entries which are actually erroneous. 
- These orders were already delivered, suggesting that they must have been approved already, yet they are missing their order_approved_at dates. Hence, these orders are erroneous. 
- However, they should not be dropped as the date of the approval should not affect the Exploratory Data Analysis or the modelling process as this data field is not critical to determining the potential for customers to return.

order_delivered_carrier_date:
- There are small proportions of null entries in the order_delivered_carrier_date (1.79%). 
- 1781 of these entries are valid and do not need to be dropped as the orders were not delivered to the carrier and therefore do not have a order_delivered_carrier_date entry.
- 2 of these entries are erroneous. These 2 orders have already been delivered to the customer by the carrier. This suggests that there should be an order_delivered_carrier_date for these 2 data entries. However, they do not exist and hence these missing values are erroneous.
- However, they are not critical to the EDA or modelling as it is not imperative to determining the potential for customers to return.

order_delivered_customer_date:
- There are small proportions of null entries in the order_delivered_carrier_date (2.98%).
- 2957 of the entries are valid as the order was not delivered to the customer and therefore it would make sense that there would be a missing order_delivered_customer__date value.
- The other 8 of these entries are erroneous. The orders were already delivered, as shown in their order_status field, however, the date in which the customer received it was not entered into the dataset in the order_delivered_customer_date field. Hence, these 8 entries are erroneous data entries.
- However, they are not critical to the EDA or modelling as they are not important in determining returning customers.

General Notes:
- Olist's dataset may be containing inaccurate or outdated information of the order statuses as there are multiple entries with the delivered statuses that do not have the order_approved_at, order_delivered_carrier_date, order_delivered_carrier_date entries despite the order being considered 'delivered'.
- Generally, the erroneous entries will not need to be dropped as they are not important to the EDA or modelling. Furthermore, the other information in the other fields will be more important to be kept to ensure the dataset is more complete for the EDA.

### Products
Investigates the missing values in the products dataset

In [48]:
# Investigate the proportion of null entries in each data field of the products dataset
calculate_null_proportion_per_column(products_df)

% of null entries per column:


product_id                    0.00
product_category_name         1.85
product_name_lenght           1.85
product_description_lenght    1.85
product_photos_qty            1.85
product_weight_g              0.01
product_length_cm             0.01
product_height_cm             0.01
product_width_cm              0.01
dtype: float64

In [49]:
# Identify the number of rows with at least 1 null value in the products dataset
count_rows_with_nulls(products_df)

No. rows with at least one null value: 611
% of rows with at least one null value: 1.85%


In [50]:
# Identify the rows with at least 1 null value
products_df[products_df.isnull().any(axis=1)]

product_id product_category_name  \
105    a41e356c76fab66334f36de622ecbd3a                   NaN   
128    d8dee61c2034d6d075997acef1870e9b                   NaN   
145    56139431d72cd51f19eb9f7dae4d1617                   NaN   
154    46b48281eb6d663ced748f324108c733                   NaN   
197    5fb61f482620cb672f5e586bb132eae9                   NaN   
...                                 ...                   ...   
32515  b0a0c5dd78e644373b199380612c350a                   NaN   
32589  10dbe0fbaa2c505123c17fdc34a63c56                   NaN   
32616  bd2ada37b58ae94cc838b9c0569fecd8                   NaN   
32772  fa51e914046aab32764c41356b9d4ea4                   NaN   
32852  c4ceee876c82b8328e9c293fa0e1989b                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
105                    NaN                         NaN                 NaN   
128                    NaN                         NaN                 NaN   
145                    NaN                         NaN                 NaN   
154                    NaN                         NaN                 NaN   
197                    NaN                         NaN                 NaN   
...                    ...                         ...                 ...   
32515                  NaN                         NaN                 NaN   
32589                  NaN                         NaN                 NaN   
32616                  NaN                         NaN                 NaN   
32772                  NaN                         NaN                 NaN   
32852                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
105               650.0               17.0               14.0   
128               300.0               16.0                7.0   
145               200.0               20.0               20.0   
154             18500.0               41.0               30.0   
197               300.0               35.0                7.0   
...                 ...                ...                ...   
32515            1800.0               30.0               20.0   
32589             800.0               30.0               10.0   
32616             200.0               21.0                8.0   
32772            1300.0               45.0               16.0   
32852             700.0               28.0                3.0   

       product_width_cm  
105                12.0  
128                20.0  
145                20.0  
154                41.0  
197                12.0  
...                 ...  
32515              70.0  
32589              23.0  
32616              16.0  
32772              45.0  
32852              43.0  

[611 rows x 9 columns]

In [51]:
# Identify the rows with missing product_category_name
products_df[products_df['product_category_name'].isnull()]

product_id product_category_name  \
105    a41e356c76fab66334f36de622ecbd3a                   NaN   
128    d8dee61c2034d6d075997acef1870e9b                   NaN   
145    56139431d72cd51f19eb9f7dae4d1617                   NaN   
154    46b48281eb6d663ced748f324108c733                   NaN   
197    5fb61f482620cb672f5e586bb132eae9                   NaN   
...                                 ...                   ...   
32515  b0a0c5dd78e644373b199380612c350a                   NaN   
32589  10dbe0fbaa2c505123c17fdc34a63c56                   NaN   
32616  bd2ada37b58ae94cc838b9c0569fecd8                   NaN   
32772  fa51e914046aab32764c41356b9d4ea4                   NaN   
32852  c4ceee876c82b8328e9c293fa0e1989b                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
105                    NaN                         NaN                 NaN   
128                    NaN                         NaN                 NaN   
145                    NaN                         NaN                 NaN   
154                    NaN                         NaN                 NaN   
197                    NaN                         NaN                 NaN   
...                    ...                         ...                 ...   
32515                  NaN                         NaN                 NaN   
32589                  NaN                         NaN                 NaN   
32616                  NaN                         NaN                 NaN   
32772                  NaN                         NaN                 NaN   
32852                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
105               650.0               17.0               14.0   
128               300.0               16.0                7.0   
145               200.0               20.0               20.0   
154             18500.0               41.0               30.0   
197               300.0               35.0                7.0   
...                 ...                ...                ...   
32515            1800.0               30.0               20.0   
32589             800.0               30.0               10.0   
32616             200.0               21.0                8.0   
32772            1300.0               45.0               16.0   
32852             700.0               28.0                3.0   

       product_width_cm  
105                12.0  
128                20.0  
145                20.0  
154                41.0  
197                12.0  
...                 ...  
32515              70.0  
32589              23.0  
32616              16.0  
32772              45.0  
32852              43.0  

[610 rows x 9 columns]

Notes:
- From briefly observing the products dataset, the null values in product_category_name, product_name_lenght, product_description_lenght, and product_photos_qty occur together. This can be seen as the proportions of null values in these fields are equal (1.85%), and the dataset above shows that null values in these 4 fields usually occur together.
- These entries are likely data entry errors as usually products will come with a name for customers to search, identify and order. Furthermore, since these missing entries occur together, it suggests a data entry issue in the system, likely due to it not being able to retrieve the information of these products.
- These will need to be dropped as product_description_lenght may be used in the EDA and modelling as the use of longer text and tags may affect online search results.
- They can be dropped as they are only 1.85% of the dataset, which is small and negligible.

In [52]:
# Identify the rows with missing product_category_name
products_df[products_df['product_weight_g'].isnull()]

product_id product_category_name  \
8578   09ff539a621711667c43eba6a3bd8466                 bebes   
18851  5eb564652db742ff8f28759cd8d2652a                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
8578                  60.0                       865.0                 3.0   
18851                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
8578                NaN                NaN                NaN   
18851               NaN                NaN                NaN   

       product_width_cm  
8578                NaN  
18851               NaN

Notes:
- The missing values in product_weight_g, product_length_cm, product_height_cm, and product_width_cm occur together as well.
- There are 2 entries which have missing values in these 4 fields.
- These are also likely data entry errors due to the system not being able to obtain information of the product as these missing values occur together, suggesting the system was unable to obtain the dimension information of the product.
- These entries do not need to be dropped as these fields are not important for EDA and modeling.

### Null Check Summary
- Drop entries with missing product_description_lenght in the products dataset
- Do not drop entries with missing values in the other datasets.

## Duplicate Check

In [80]:
# Calculates the number and proportion of rows which are duplicate entries
def duplicate_entries_search(df):
    # Identify total number of rows in the dataset
    total_rows = df.shape[0]

    # Count duplicate rows
    num_duplicate_rows = df.duplicated().sum()
    percent_duplicate_rows = (num_duplicate_rows / total_rows) * 100

    # Display results
    print(f"Total duplicate rows: {num_duplicate_rows}")
    print(f"% of duplicate rows: {percent_duplicate_rows:.2f}%")


# Calculates the number and proportion of rows which have duplicate values in a specified data field
def duplicate_values_search(df, field):
    # Identify total number of rows in the dataset
    total_rows = df.shape[0]

    # Count duplicate values in the data field
    num_duplicate_values = df[field].duplicated().sum()
    proportion_duplicate_values = (num_duplicate_values / total_rows) * 100

    # Display results
    print(f"\n{field}:")
    print(f"Total duplicate entries: {num_duplicate_values}")
    print(f"% of duplicate entries: {proportion_duplicate_values:.2f}%")


In [112]:
# Identify the duplicate entries in the customers dataset
duplicate_entries_search(customers_df)
duplicate_values_search(customers_df, 'customer_id')

Total duplicate rows: 0
% of duplicate rows: 0.00%

customer_id:
Total duplicate entries: 0
% of duplicate entries: 0.00%


Customers dataset:
- There are no duplicate data in the customers dataset.

In [55]:
# Identify the duplicate entries in the geolocation dataset
duplicate_entries_search(geolocation_df)

Total duplicate rows: 261831
% of duplicate rows: 26.18%


In [115]:
# Observe the duplicate rows in the geolocations dataset
geolocation_df[geolocation_df.duplicated()]

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
15                            01046       -23.546081       -46.644820   
44                            01046       -23.546081       -46.644820   
65                            01046       -23.546081       -46.644820   
66                            01009       -23.546935       -46.636588   
67                            01046       -23.546081       -46.644820   
...                             ...              ...              ...   
1000153                       99970       -28.343273       -51.873734   
1000154                       99950       -28.070493       -52.011342   
1000159                       99900       -27.877125       -52.224882   
1000160                       99950       -28.071855       -52.014716   
1000162                       99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
15             sao paulo                SP  
44             sao paulo                SP  
65             sao paulo                SP  
66             sao paulo                SP  
67             sao paulo                SP  
...                  ...               ...  
1000153          ciriaco                RS  
1000154         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000162         tapejara                RS  

[261831 rows x 5 columns]

Customers dataset:
- 26.18% of the geolocations entries are duplicate.
- They will need to be dropped as they are unnecessary. The duplicate entries will cause downstream merges to contain more duplicate entries as the merge may explode the geolocation dataset with the duplicate entries. Hence, they need to be dropped.

In [87]:
# Identify the duplicate entries in the items dataset
duplicate_entries_search(items_df)

Total duplicate rows: 0
% of duplicate rows: 0.00%


Items Dataset:
- No duplicate data entries.

In [99]:
# Identify the duplicate entries in the payments dataset
duplicate_entries_search(payments_df)

Total duplicate rows: 0
% of duplicate rows: 0.00%


Payments Dataset:
- No duplicate data entries.

In [102]:
# Identify the duplicate entries in the reviews dataset
duplicate_entries_search(reviews_df)
duplicate_values_search(reviews_df, 'review_id')
duplicate_values_search(reviews_df, 'order_id')


Total duplicate rows: 0
% of duplicate rows: 0.00%

review_id:
Total duplicate entries: 814
% of duplicate entries: 0.82%

order_id:
Total duplicate entries: 551
% of duplicate entries: 0.56%


Reviews Dataset:
- s

In [104]:
# Identify the duplicate entries in the orders dataset
duplicate_entries_search(orders_df)
duplicate_values_search(orders_df, 'order_id')

Total duplicate rows: 0
% of duplicate rows: 0.00%

order_id:
Total duplicate entries: 0
% of duplicate entries: 0.00%


Orders Dataset:
- No duplicate data entries.

In [106]:
# Identify the duplicate entries in the products dataset
duplicate_entries_search(products_df)
duplicate_values_search(products_df, 'product_id')

Total duplicate rows: 0
% of duplicate rows: 0.00%

product_id:
Total duplicate entries: 0
% of duplicate entries: 0.00%


In [108]:
# Identify the duplicate entries in the sellers dataset
duplicate_entries_search(sellers_df)
duplicate_values_search(sellers_df, 'seller_id')

Total duplicate rows: 0
% of duplicate rows: 0.00%

seller_id:
Total duplicate entries: 0
% of duplicate entries: 0.00%


Sellers Dataset:
- No duplicate data entries.

In [110]:
# Identify the duplicate entries in the translations dataset
duplicate_entries_search(translations_df)
duplicate_values_search(translations_df, 'product_category_name')
duplicate_values_search(translations_df, 'product_category_name_english')

Total duplicate rows: 0
% of duplicate rows: 0.00%

product_category_name:
Total duplicate entries: 0
% of duplicate entries: 0.00%

product_category_name_english:
Total duplicate entries: 0
% of duplicate entries: 0.00%


Translations Dataset:
- No duplicate data entries.

## Geolocation Check
Checks the geolocation which is not Brazil.

## Payment Sequential Check
Investigate the outlier in the payment_sequential data field in the payments dataset.

## Product Weight Check
Investigate the low outlier value in the product weight field in the products dataset.